app/src/main/java/org/simple/clinic/appupdate/CheckAppUpdateAvailability.kt

```diff
@@ -0,0 +1,65 @@
+package org.simple.clinic.appupdate
+
+import android.app.Application
+import android.os.Build
+import androidx.annotation.VisibleForTesting
+import com.google.android.play.core.appupdate.AppUpdateInfo
+import com.google.android.play.core.appupdate.AppUpdateManagerFactory
+import com.google.android.play.core.install.model.AppUpdateType.FLEXIBLE
+import com.google.android.play.core.install.model.UpdateAvailability
+import io.reactivex.Single
+import org.simple.clinic.util.None
+import org.simple.clinic.util.Optional
+import org.simple.clinic.util.toOptional
+import timber.log.Timber
+import javax.inject.Inject
+
+class CheckAppUpdateAvailability @Inject constructor(
+    private val appContext: Application,
+    private val config: AppUpdateConfig,
+    private val versionUpdateCheck: (Int, Application, AppUpdateConfig) -> Boolean = isVersionApplicableForUpdate
+) {
+
+  fun listen(): Single<Optional<AppUpdateData>> {
+    val appUpdateManager = AppUpdateManagerFactory.create(appContext)
+    val appUpdateInfoTask = appUpdateManager.appUpdateInfo
+
+    return Single.create { emitter ->
+
+      appUpdateInfoTask.addOnSuccessListener { appUpdateInfo ->
+        if (shouldShowUpdateDialog(appUpdateInfo)) {
+          emitter.onSuccess(AppUpdateData(appUpdateInfo).toOptional())
+        } else {
+          emitter.onSuccess(None)
+        }
+      }
+
+      appUpdateInfoTask.addOnFailureListener { exception ->
+        emitter.onError(exception)
+      }
+    }
+  }
+
+  @VisibleForTesting
+  fun shouldShowUpdateDialog(appUpdateInfo: AppUpdateInfo): Boolean {
+    return config.inAppUpdateEnabled
+        && appUpdateInfo.updateAvailability() == UpdateAvailability.UPDATE_AVAILABLE
+        && appUpdateInfo.isUpdateTypeAllowed(FLEXIBLE)
+        && versionUpdateCheck(appUpdateInfo.availableVersionCode(), appContext, config)
+  }
+}
+
+private val isVersionApplicableForUpdate = { availableVersionCode: Int, appContext: Application, config: AppUpdateConfig ->
```

Comment:
This can be a function too.

```kotlin
class CheckAppUpdateAvailability @Inject constructor(
    private val appContext: Application,
    private val config: AppUpdateConfig,
    private val versionUpdateCheck: (Int, Application, AppUpdateConfig) -> Boolean = ::isVersionApplicableForUpdate
) {
}

private fun isVersionApplicableForUpdate(availableVersionCode: Int, appContext: Application, config: AppUpdateConfig): Boolean {
  val packageManager = appContext.packageManager.getPackageInfo(appContext.packageName, 0)

  return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
    val longVersionCode = packageManager.longVersionCode
    val versionCode = longVersionCode.and(0xffffffff)
    availableVersionCode.minus(versionCode) >= config.numberOfVersionsToDiff

  } else {
    availableVersionCode.minus(packageManager.versionCode) >= config.numberOfVersionsToDiff
  }
}
```

I'm not asking for this to be changed, just adding that this is also an option instead of needing to declare a lambda.

Reply:
Cool. I will make a note of this.